pipit/readers/nsight_reader.py

```diff
@@ -0,0 +1,129 @@
+# Copyright 2022 Parallel Software and Systems Group, University of Maryland.
+# See the top-level LICENSE file for details.
+#
+# SPDX-License-Identifier: MIT
+
+import pandas as pd
+import pipit.trace
+import pipit.graph as graph
+
+
+class NSightReader:
+    """Reader for NSight trace files"""
+
+    def __init__(self, file_name) -> None:
+        self.file_name = file_name
+        self.df = None
+
+    """
+    This read function directly takes in a csv of the trace report and
+    utilizes pandas to convert it from a csv into a dataframe.
+    """
+
+    def read(self):
+        # Read in csv
+        self.df = pd.read_csv(self.file_name)
+
+        graph = self.call_graph()
+
+        # Copy data into new dataframe
+        df = self.df
+        df2 = df.copy()
+
+        # Create new columns for df with start time
+        df["Event Type"] = "Enter"
+        df["Time"] = df["Start (ns)"]
+        df["Graph_Node"] = None
+        df["Level"] = None
+
+        for i in range(len(df)):
+            node = graph.get_node(df.iloc[i]["RangeStack"])
+            df.at[i, "Graph_Node"] = node
+            df.at[i, "Level"] = node.get_level()
+
+        # Create new columns for df2 with end time
+        df2["Event Type"] = "Exit"
+        df2["Time"] = df2["End (ns)"]
+
+        for i in range(len(df2)):
+            node = graph.get_node(df2.iloc[i]["RangeStack"])
+            df2.at[i, "Graph_Node"] = node
+            df2.at[i, "Level"] = node.get_level()
+
+        df2.Level = df2.Level.astype(int)
+
+        # Combine dataframes together
+        df = pd.concat([df, df2])
+
+        # Tidy Dataframe
+        df.drop(["Start (ns)", "End (ns)"], axis=1, inplace=True)
+
+        df.sort_values(by="Time", ascending=True, inplace=True)
+
+        df.reset_index(drop=True, inplace=True)
+
+        return pipit.trace.Trace(None, df)
+
+    """
+    This call graph function takes in a dataframe and shows the user
+    how the call graph would look like within their program.
+    """
+
+    def call_graph(self):
+        df = self.df
+
+        # Creating call graph
+        call_graph = graph.Graph()
+
+        # Creating a stack
+        stack = []
+
+        for i in range(len(df)):
```

Comment:
Would we need some kind of filtering by rank/thread here before we iterate over the rows to create the cct?

Reply:
I think so. I don't have inputs to test with. Will keep it in mind.